e3sm_diags/driver/utils/dataset_xr.py

```diff
@@ -1172,38 +1172,32 @@ def _get_time_slice(self, ds: xr.Dataset, filepaths: List[str]) -> slice:
 
         return slice(start_time, end_time)
 
-    def _parse_years_from_filepaths(self, filepaths: List[str]) -> Tuple[int, int]:
-        """Parse the start and end years from the filepaths.
+    def _extract_var_start_and_end_years(self, ds: xr.Dataset) -> Tuple[int, int]:
+        """Extract the start and end years from the time coordinates.
 
-        If there are more than one file, the start and end years are parsed
-        from the first and last files.
+        If the last time coordinate starts in January, subtract one year from
+        the end year to get the correct end year which should align with the
+        end year from the filepaths.
 
         Parameters
         ----------
-        filepaths : List[str]
-            The list of filepaths.
+        ds : xr.Dataset
+            The dataset with time coordinates.
 
         Returns
         -------
         Tuple[int, int]
             The start and end years.
         """
-        if len(filepaths) > 1:
-            # Example: ../TS_198501_198612.nc' and ../TS_198701_198812.nc'
-            start_file = filepaths[0]
-            end_file = filepaths[-1]
-
-            # 1985 and 1988
-            var_start_year = int(start_file.split("_")[-2][:4])
-            var_end_year = int(end_file.split("_")[-1][:4])
-        else:
-            # Example: {var}_{start_yr}01_{end_yr}12.nc
-            filename_to_parse = filepaths[0]
+        time_dim = xc.get_dim_keys(ds, axis="T")
+        time_coords = ds.indexes[time_dim].to_datetimeindex()
+        start_year = time_coords[0].year
+        end_year = time_coords[-1].year
 
-            var_start_year = int(filename_to_parse.split("/")[-1].split("_")[-2][:4])
-            var_end_year = int(filename_to_parse.split("/")[-1].split("_")[-1][:4])
+        if time_coords[-1].month == 1:
+            end_year -= 1
 
-        return var_start_year, var_end_year
+        return start_year, end_year
```

Comment:
@chengzhuzhang `_extract_var_start_and_end_years()` replaces `_parse_years_from_filepaths()`.

It uses the time coordinates to get the start and end years, rather than filepaths. If the last time coordinate extends to the next year (`month == 1`), then it subtracts 1 from the end year.

For example: 
* (2000-01-01, 2001-12-01) -> 2000 and 2001
* (2000-01-01, 2002-01-01) -> 2000 and 2001 (subtract 1 from 2001)